### PR TITLE
Error-chain -> failure

### PR DESCRIPTION
**Pull request by steveklabnik**
This PR ports us from error-chain to @withoutboats' `failure` crate.

I'm done for the day so there's probably CI failures and such, but I wanted to toss this up there so that everyone can see.

/cc @aturon @alexcrichton

### CODE REVIEW

**Comment by mgattozzi** [comment]
@steveklabnik curious is there a reason to move towards that over error-chain?

File: src/lib.rs [hunk 1]
```diff
@@ -172,9 +181,9 @@ pub fn build(config: &Config, artifacts: &[&str]) -> Result<()> {
             .stderr(Stdio::piped())
             .spawn()
             .map_err(|e| if e.kind() == io::ErrorKind::NotFound {
-                Error::with_chain(e, format!("frontend `{}` not found", frontend))
+                format_err!("frontend `{}` not found", frontend).into()
```
Comment (euclio) [inline src/lib.rs]: It looks like you could use `chain()` to keep the context here?

File: src/lib.rs [hunk 2]
```diff
@@ -196,8 +205,8 @@ pub fn build(config: &Config, artifacts: &[&str]) -> Result<()> {
 
 /// Run all documentation tests.
 pub fn test(config: &Config) -> Result<()> {
-    let doc_json = File::open(config.documentation_path()).chain_err(
-        || "could not find generated documentation",
+    let doc_json = File::open(config.documentation_path()).map_err(|_|
```
Comment (euclio) [inline src/lib.rs]: Same here.

**Comment by euclio** [comment]
This looks interesting!

**Comment by steveklabnik** [comment]
> curious is there a reason to move towards that over error-chain?

It's speculated that this kind of library may be preferable to error-chain overall; I wanted to give it a shot to see what the diff looks like. I think I personally prefer it, but the only way to find out is to give it a try.

**Comment by mgattozzi** [comment]
Awesome!

File: src/lib.rs (hunk 1 above)
Comment (steveklabnik) [reply src/lib.rs]: wait, what's chain()? I don't see it in failure's docs

Comment (euclio) [reply src/lib.rs]: Looks like it's been renamed to `context`? It works slightly differently that `chain_err` did in `error_chain`.

Comment (withoutboats) [reply src/lib.rs]: Yea I renamed it to context. You could do `e.context(format!(...)).into()` here.

**Comment by steveklabnik** [comment]
@withoutboats while working on the test part of this PR, 

```
    = note: expected type `std::result::Result<_, failure::Error>`
               found type `std::result::Result<_, failure::Context<InvalidDirective>>`
```

Do you have to do something specific to use `Context`? It doesn't implement `Error`?

**Comment by withoutboats** [comment]
`Context` implements `Fail`; to get an `Error` all `Fail`ures have to be cast using `?` or `into()`.

**Comment by steveklabnik** [comment]
Thanks boats!

Now our test is failing because of the way we test the context; I suspect that I have to update the macro in some way. More work needed.

**Comment by euclio** [comment]
In this case, it might just be better to check that the cause is correct and always return an `InvalidDirective` failure from `parse_test`?

**Comment by steveklabnik** [comment]
I've rebased this, and it compiles, but has test failures about not being able to find the source test crates. Hrm.

**Comment by steveklabnik** [comment]
Ohh interesting! https://travis-ci.org/steveklabnik/rustdoc/jobs/305812561

even though https://github.com/steveklabnik/rustdoc/pull/203 passed, it's now failing. So it's likely *that* PR, not this one.... @euclio any ideas?

**Comment by euclio** [comment]
Failing on master too. Probably related to a change in how the analysis is generated. I did a quick test and it should be fixable by updating `rls-analysis` and `rls-data`, though there's an unrelated failure in `structs.rs` as well.

**Comment by steveklabnik** [comment]
ah, that makes perfect sense.

ill dig in.

**Comment by steveklabnik** [comment]
Updated and rebased, still failing. I must have missed something somewhere...

**Comment by steveklabnik** [comment]
oh wait looks like a bad rebase....

**Comment by steveklabnik** [comment]
Got master passing, this seems to reintroduce it. Hmmmmmmmmm

**Comment by steveklabnik** [comment]
@mgattozzi @euclio okay I think this is ready for a review, maybe you'll spot where I messed up. @withoutboats if you have a second as well, that'd be nice.

File: tests/source.rs [hunk 3]
```diff
@@ -401,37 +423,49 @@ fn parse_test(line: &str) -> Option<Result<TestCase>> {
...
-            let regex = Regex::new(&args[1]).chain_err(|| "invalid regex syntax")?;
+            let regex = Regex::new(&args[1]).map_err(|e| {
+                failure::Error::from(e.context("invalid regex syntax"))
+            })?;
 
             Directive::Has(regex)
         }
         "matches" => {
-            ensure!(args.len() == 2, "not enough arguments");
+            if args.len() == 2 {
```
Comment (euclio) [inline tests/source.rs]: This should be `if args.len() != 2`

File: tests/source.rs [hunk 4]
```diff
@@ -401,37 +423,49 @@ fn parse_test(line: &str) -> Option<Result<TestCase>> {
...
-            )?;
+            let value = serde_json::from_str(&args[1]).map_err(|e| {
+                failure::Error::from(e.context("invalid JSON syntax"))
+            })?;
 
             Directive::Matches(value)
         }
         "assert" => {
-            ensure!(args.len() == 1, "expected 1 argument");
+            if args.len() == 1 {
```
Comment (euclio) [inline tests/source.rs]: This should be `if args.len() != 1`

File: tests/source.rs (hunk 3 above)
Comment (steveklabnik) [reply tests/source.rs]: lol oops

**Comment by steveklabnik** [comment]
Awesome, thank you! Of course I messed that up... that leads to just one failure now.

File: tests/source.rs [hunk 5]
```diff
@@ -391,7 +413,7 @@ fn parse_test(line: &str) -> Option<Result<TestCase>> {
     if let Some(caps) = DIRECTIVE_RE.captures(line) {
         let directive = &caps["directive"];
         let result = parse_directive(directive, &caps["args"], caps.name("negated").is_some())
-            .chain_err(|| ErrorKind::InvalidDirective(line.into()));
+            .map_err(|e| e.context(InvalidDirective { d: line.into() }).into());
```
Comment (euclio) [inline tests/source.rs]: I think the context is backwards here. I think that `parse_test` should return `Option<Result<TestCase, InvalidDirective>>`. Make sure to set the context of the `InvalidDirective` error to be the original error. The tests should then check that `err.cause()` is the expected type.

Comment (steveklabnik) [reply tests/source.rs]: yeah, this is sorta what i'm thinking as well. been toying with this...

File: tests/source.rs [hunk 6]
```diff
@@ -401,37 +423,49 @@ fn parse_test(line: &str) -> Option<Result<TestCase>> {
 fn parse_directive(directive: &str, args: &str, negated: bool) -> Result<TestCase> {
     let args = match shlex::split(args) {
```
Comment (euclio) [inline tests/source.rs]: This can be `ok_or_else`

**Comment by euclio** [comment]
Looks good with comments addressed. Any reason you're preferring `ok_or` over `ok_or_else`? I don't think clippy likes that.

**Comment by steveklabnik** [comment]
> Any reason you're preferring ok_or over ok_or_else? I don't think clippy likes that.

no reason; I'll run clippy too. 

now dealing with this fun error:

```
        thread 'tests::parse_test' panicked at 'unexpected error kind, got InvalidDirective { d: "// @has /test \'[\'" }

invalid JMESPath syntax but expected "InvalidDirective"', tests\source.rs:583:8
```

hmmmmmmmmmmmmmmmmm

**Comment by euclio** [comment]
Yeah, I was confused by that too. Does downcasting the cause to `jmespath::JmespathError` work?

**Comment by steveklabnik** [comment]
> Yeah, I was confused by that too. Does downcasting the cause to jmespath::JmespathError work?

It doesn't seem to, no matter what I try to cast it to, it doesn't seem to do the right thing. Hrm.

**Comment by steveklabnik** [comment]
@euclio so check out this last commit: if we do make it always return `InvalidDirective`, then there's no reason to assert that type about it. this is now doing something similar to https://boats.gitlab.io/failure/error-errorkind.html; maybe we should do that exactly?

**Comment by euclio** [comment]
Exactly. Error/ErrorKind could work, but since this is an internal-only error that complexity might not be warranted. We could just make it an enum like [here](https://boats.gitlab.io/failure/custom-fail.html#caveats-on-this-pattern).

**Comment by steveklabnik** [comment]
Time to :shipit: , if we want to tweak that last error, we can, but I don't think it matters a ton, honestly.

**Pull request merged**